### PR TITLE
Week10 and Week 11

### PR DESCRIPTION
**Pull request by hvariant**
doing week11 helped me find and fix a bug in week10, and they are connected anyways, so...

### CODE REVIEW

File: cis194/week10/zhansongl/SExpr.hs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+parseAtom :: Parser Atom
+parseAtom = spaces *> atom <* spaces
+  where atom = (I <$> ident) <|> (N <$> posInt)
+
+parseSExpr :: Parser SExpr
+parseSExpr = A <$> parseAtom
+             <|>
+             openParenthesis *> (Comb <$> oneOrMore parseSExpr) <* closeParenthesis
+  where openParenthesis = spaces *> (char '(')
+        closeParenthesis = (char ')') <* spaces
```
Comment (stevemao) [inline cis194/week10/zhansongl/SExpr.hs]: parentheses are not needed :)

Comment (hvariant) [reply cis194/week10/zhansongl/SExpr.hs]: why not?

Comment (stevemao) [reply cis194/week10/zhansongl/SExpr.hs]: function call has higher precedence than `<*`

Comment (stevemao) [reply cis194/week10/zhansongl/SExpr.hs]:
```suggestion
        closeParenthesis = (char ')') <* spaces
        closeParenthesis = char ')' <* spaces
```

Comment (hvariant) [reply cis194/week10/zhansongl/SExpr.hs]: oh

File: cis194/week10/zhansongl/AParser.hs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+parserFmap :: (a -> b) -> Parser a -> String -> Maybe (b, String)
+parserFmap f pa s = first f <$> runParser pa s 
+
+parserReplace :: a -> Parser b -> String -> Maybe (a, String)
+parserReplace a pb s = first (const a) <$> runParser pb s
+
+instance Functor Parser where
+  fmap f = Parser . parserFmap f
+  (<$) a = Parser . parserReplace a
```
Comment (cht8687) [inline cis194/week10/zhansongl/AParser.hs]: why not:

instance Functor Parser where
  fmap f (Paser g) = Parser $ fmap (first f) . g

Comment (hvariant) [reply cis194/week10/zhansongl/AParser.hs]: hmm, good catch, will fix tomorrow

Comment (cht8687) [reply cis194/week10/zhansongl/AParser.hs]: 😄 I copied from https://github.com/fp-works/2019-winter-Haskell-school/blob/master/cis194/week5/robert-chang/src/Parser.hs#L23

Comment (hvariant) [reply cis194/week10/zhansongl/AParser.hs]: done 😎

File: cis194/week10/zhansongl/AParser.hs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+instance Functor Parser where
+  fmap f (Parser g) = Parser $ fmap (first f) . g
+  (<$) a (Parser b) = Parser $ fmap (first $ const a) . b
+
+instance Applicative Parser where
+  pure a = Parser $ Just . ((,) a)
+  (<*>) (Parser pf) (Parser pa) = Parser $ \s -> pf s >>= \(f, s') -> first f <$> pa s'
+
+abParser :: Parser (Char, Char)
+abParser = (,) <$> (char 'a') <*> (char 'b')
```
Comment (shineli1984) [inline cis194/week10/zhansongl/AParser.hs]: nit: 
function application named the space between a function and its argument has the highest precedence.
```suggestion
abParser = (,) <$> char 'a' <*> char 'b'
```

File: cis194/week10/zhansongl/AParser.hs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+intPair :: Parser [Integer]
+intPair = (\x y -> [x, y]) <$> posInt <* char ' ' <*> posInt
+
+instance Alternative Parser where
+  empty = Parser $ const Nothing
+  (<|>) (Parser a) (Parser a') = Parser $ \s -> a s <|> a' s
+
+intOrUppercase :: Parser ()
+intOrUppercase = () <$ posInt <|> () <$ satisfy isUpper
```
Comment (shineli1984) [inline cis194/week10/zhansongl/AParser.hs]: You may be interested in a function `void :: Functor f => f a -> f ()`, that can make this implementation much cleaner.

**Pull request merged**